pkg/client/s3client.go

```diff
@@ -235,41 +238,101 @@ func (c *S3Client) CopyObjectToBucket(srcBucketName string, destBucketName strin
 }
 
 type CustomReader struct {
-	fp      *os.File
-	size    int64
-	read    int64
-	signMap map[int64]struct{}
-	mux     sync.Mutex
+	fp              *os.File
+	size            int64
+	read            int64
+	signMap         map[int64]struct{}
+	mux             sync.Mutex
+	progresstracker *ProgressTracker
 }
 
-func (r *CustomReader) Read(p []byte) (int, error) {
-	return r.fp.Read(p)
+type ProgressTracker struct {
+	progress *mpb.Progress
+	bar      *mpb.Bar
+	isBarSet bool
+	counter  *formattedCounter
 }
 
-func (r *CustomReader) ReadAt(p []byte, off int64) (int, error) {
+type formattedCounter struct {
+	read  *int64
+	total int64
+}
+
+func (f *formattedCounter) Decor(stat decor.Statistics) (str string, viewWidth int) {
+	str = fmt.Sprintf("%s / %s", formatBytes(*f.read), formatBytes(f.total))
+	return str, len(str)
+}
+
+func (f *formattedCounter) Format(string) (_ string, width int) {
+	return "", 0
+}
+
+func (f *formattedCounter) Sync() (chan int, bool) {
+	return nil, false
+}
+
+func (r *CustomReader) Read(p []byte) (n int, err error) {
+	n, err = r.fp.Read(p)
+	if err == nil {
+		r.read += int64(n)
+		if r.read > r.size {
+			r.read = r.size
+		}
+		r.progresstracker.counter.read = &r.read
+		r.progresstracker.bar.SetCurrent(r.read)
+	}
+	return n, err
+}
+
+func (r *CustomReader) ReadAt(p []byte, off int64, tracker *ProgressTracker) (int, error) {
+	r.mux.Lock()
+	klog.Infof("readat")
+	defer r.mux.Unlock()
 	n, err := r.fp.ReadAt(p, off)
 	if err != nil {
+		if err == io.EOF {
+			return n, nil
+		}
 		return n, err
 	}
-	r.mux.Lock()
+	r.read += int64(n)
+	r.progresstracker.bar.SetCurrent(r.read)
 	if _, ok := r.signMap[off]; ok {
-		r.read += int64(n)
 		progress := int(float32(r.read*100) / float32(r.size))
-		fmt.Printf("\rUploading: Total read(bytes):%d    progress:%d%%", r.read, progress)
+		klog.Infof("\rUploading: Total read(bytes): %d    progress: %d%%", r.read, progress)
```

Comment:
I'm unsure if we'd be needing this anymore as we're tracking the upload through the progress bar.